bindings/rust/wasmedge-sys/src/instance/function.rs

```diff
@@ -0,0 +1,162 @@
+    let raw_returns = unsafe { std::slice::from_raw_parts_mut(returns, return_len) };
+
+    HOST_FUNCS.with(|f| {
+        let mut host_functions = f.borrow_mut();
+        let real_fn = host_functions
+            .remove(&key)
+            .expect("host function should there");
+        result = real_fn(input);
+    });
+
+    match result {
+        Ok(v) => {
+            assert!(v.len() == return_len);
+            for (idx, item) in v.into_iter().enumerate() {
+                raw_returns[idx] = item.into();
+            }
+            wasmedge::WasmEdge_Result { Code: 0 }
+        }
+        Err(c) => wasmedge::WasmEdge_Result { Code: c as u8 },
+    }
+}
+
+#[derive(Debug)]
+pub struct Function {
+    pub(crate) ctx: *mut wasmedge::WasmEdge_FunctionInstanceContext,
+    pub(crate) registed: bool,
+    ty: Option<Type>,
+}
+
+impl Function {
+    /// wasmedge::WasmEdge_FunctionInstanceCreate take C functions
+    /// This may not be implement
+    pub fn create<I: WasmFnIO, O: WasmFnIO>(
+        _f: Box<dyn std::ops::Fn(Vec<Value>) -> Vec<Value>>,
+    ) -> Self {
+        unimplemented!()
+    }
+
+    // TODO:
+    // binding errors and restict the error types
+    #[allow(clippy::type_complexity)]
+    /// Binding Rust function (HostFunction) to WasmEdgeFunction
+    pub fn create_bindings<I: WasmFnIO, O: WasmFnIO>(
+        real_fn: Box<dyn Fn(Vec<Value>) -> Result<Vec<Value>, u8>>,
+    ) -> Self {
+        let mut key = 0usize;
+        HOST_FUNCS.with(|f| {
+            let mut rng = rand::thread_rng();
+            let mut host_functions = f.borrow_mut();
+            assert!(
+                host_functions.len() <= host_functions.capacity(),
+                "Too many host functions"
+            );
+            key = rng.gen::<usize>();
+            while host_functions.contains_key(&key) {
+                key = rng.gen::<usize>();
+            }
+            host_functions.insert(key, real_fn);
+        });
+
+        let key_ptr = key as *const usize as *mut c_void;
+        let ty = Type::create(I::parameters(), O::parameters());
+
+        let ctx = unsafe {
+            wasmedge::WasmEdge_FunctionInstanceCreateBinding(
+                ty.ctx,
+                Some(wraper_fn),
+                key_ptr,
+                std::ptr::null_mut(),
+                0,
+            )
+        };
+        Self {
+            ctx,
+            ty: Some(ty),
+            registed: false,
+        }
+    }
+}
+
+impl Drop for Function {
+    fn drop(&mut self) {
+        self.ty = None;
+        if !self.registed {
+            unsafe { wasmedge::WasmEdge_FunctionInstanceDelete(self.ctx) };
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Type {
+    pub(crate) ctx: *mut wasmedge::WasmEdge_FunctionTypeContext,
+}
+
+impl Type {
+    pub(crate) fn create(input: Vec<Value>, output: Vec<Value>) -> Self {
+        let raw_input = {
+            let mut _a = vec![wasmedge::WasmEdge_ValType_ExternRef];
```

Comment:
Can we have a meaningful name to replace `_a` and `_b`.

Reply:
Okay, I will change these into `head` and `tail`.